File: packages/transport-acp-stdio/src/acp-client-handler.ts
import * as fs from 'node:fs/promises';
import { spawn, type ChildProcess } from 'node:child_process';
import { randomUUID } from 'node:crypto';
import type { Client, RequestPermissionRequest, RequestPermissionResponse, SessionNotification } from '@agentclientprotocol/sdk/dist/acp.js';
import { RequestError } from '@agentclientprotocol/sdk/dist/acp.js';
import type { AgentEvent } from '@acme/shared-types';
import { mapSessionUpdate } from './session-update-mapper.js';
import { PermissionResolver } from './permission-resolver.js';

type Emitter = { emit(event: AgentEvent): void };

interface TerminalEntry {
  process: ChildProcess;
  output: string;
  exitCode: number | null;
  exitSignal: string | null;
  exitPromise: Promise<void>;
  exitResolve: () => void;
}

export class AcpClientHandler implements Client {
  private readonly terminals = new Map<string, TerminalEntry>();

  constructor(
    private readonly emitter: Emitter,
    readonly permissionResolver: PermissionResolver,
  ) {}

  async sessionUpdate(params: SessionNotification): Promise<void> {
    const events = mapSessionUpdate(params.sessionId, params.update);
    for (const event of events) {
      this.emitter.emit(event);
    }
  }

  async requestPermission(params: RequestPermissionRequest): Promise<RequestPermissionResponse> {
    return this.permissionResolver.request(params, this.emitter);
  }

  async readTextFile(params: { path: string; sessionId: string }): Promise<{ content: string }> {
    try {
      const content = await fs.readFile(params.path, 'utf-8');
      return { content };
    } catch {
      throw RequestError.resourceNotFound(params.path);
    }
  }

  async writeTextFile(params: { path: string; content: string; sessionId: string }): Promise<Record<string, never>> {
    await fs.writeFile(params.path, params.content, 'utf-8');
    return {};
  }

  async createTerminal(params: { command: string; args?: string[]; cwd?: string; sessionId: string }): Promise<{ terminalId: string }> {
    const terminalId = randomUUID();
    let exitResolve!: () => void;
    const exitPromise = new Promise<void>((resolve) => { exitResolve = resolve; });

    const proc = spawn(params.command, params.args ?? [], {
      cwd: params.cwd,
      stdio: ['ignore', 'pipe', 'pipe'],
      shell: true,
    });

    const entry: TerminalEntry = {
      process: proc,
      output: '',
      exitCode: null,
      exitSignal: null,
      exitPromise,
      exitResolve,
    };

    proc.stdout?.on('data', (chunk: Buffer) => { entry.output += chunk.toString(); });
    proc.stderr?.on('data', (chunk: Buffer) => { entry.output += chunk.toString(); });
    proc.on('exit', (code, signal) => {
      entry.exitCode = code;
      entry.exitSignal = signal;
      exitResolve();
    });

    this.terminals.set(terminalId, entry);
    return { terminalId };
  }

  async terminalOutput(params: { terminalId: string; sessionId: string }): Promise<{ output: string; truncated: boolean; exitStatus?: { exitCode?: number; signal?: string } }> {
    const entry = this.terminals.get(params.terminalId);
    if (!entry) throw RequestError.resourceNotFound(params.terminalId);
    const result: { output: string; truncated: boolean; exitStatus?: { exitCode?: number; signal?: string } } = {
      output: entry.output,
      truncated: false,
    };
    if (entry.exitCode !== null || entry.exitSignal !== null) {
      result.exitStatus = {
        ...(entry.exitCode !== null && { exitCode: entry.exitCode }),
        ...(entry.exitSignal !== null && { signal: entry.exitSignal }),
      };
    }
    return result;
  }

  async waitForTerminalExit(params: { terminalId: string; sessionId: string }): Promise<{ exitCode?: number; signal?: string }> {
    const entry = this.terminals.get(params.terminalId);
    if (!entry) throw RequestError.resourceNotFound(params.terminalId);
    await entry.exitPromise;
    return {
      ...(entry.exitCode !== null && { exitCode: entry.exitCode }),
      ...(entry.exitSignal !== null && { signal: entry.exitSignal }),
    };
  }

  async killTerminal(params: { terminalId: string; sessionId: string }): Promise<Record<string, never>> {
    const entry = this.terminals.get(params.terminalId);
    if (!entry) throw RequestError.resourceNotFound(params.terminalId);
    entry.process.kill();
    return {};
  }

  async releaseTerminal(params: { terminalId: string; sessionId: string }): Promise<Record<string, never>> {
    const entry = this.terminals.get(params.terminalId);
    if (entry) {
      entry.process.kill();
      this.terminals.delete(params.terminalId);
    }
    return {};
  }

  releaseAll(): void {
    for (const [id, entry] of this.terminals) {
      entry.process.kill();
      this.terminals.delete(id);
    }
  }
}

File: crates/aimote-backend/src/acp_client_handler.rs
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::Mutex;

use agent_client_protocol::{
    Client, RequestPermissionRequest, RequestPermissionResponse, SessionNotification,
    CreateTerminalRequest, CreateTerminalResponse, TerminalExitStatus,
    TerminalOutputRequest, TerminalOutputResponse,
    WaitForTerminalExitRequest, WaitForTerminalExitResponse,
    KillTerminalRequest, KillTerminalResponse,
    ReleaseTerminalRequest, ReleaseTerminalResponse,
    ReadTextFileRequest, ReadTextFileResponse,
    WriteTextFileRequest, WriteTextFileResponse,
    Error as AcpError, Result as AcpResult,
};
use tokio::io::AsyncReadExt;

use crate::event_sink::EventSink;
use crate::permission_resolver::PermissionResolver;
use crate::session_update_mapper::map_session_update;

struct TerminalEntry {
    child: tokio::process::Child,
    output: Arc<Mutex<String>>,
    exit_result: Arc<Mutex<Option<ExitResult>>>,
}

#[derive(Clone)]
struct ExitResult {
    code: Option<i32>,
}

pub struct AcpClientHandler {
    sink: Arc<dyn EventSink>,
    permission_resolver: Arc<Mutex<PermissionResolver>>,
    terminals: Mutex<HashMap<String, TerminalEntry>>,
}

impl AcpClientHandler {
    pub fn new(
        sink: Arc<dyn EventSink>,
        permission_resolver: Arc<Mutex<PermissionResolver>>,
    ) -> Self {
        Self {
            sink,
            permission_resolver,
            terminals: Mutex::new(HashMap::new()),
        }
    }

    pub async fn release_all(&self) {
        let mut terminals = self.terminals.lock().await;
        for (_, mut entry) in terminals.drain() {
            let _ = entry.child.kill().await;
        }
    }
}

#[async_trait::async_trait(?Send)]
impl Client for AcpClientHandler {
    async fn session_notification(&self, args: SessionNotification) -> AcpResult<()> {
        let session_id = args.session_id.to_string();
        let events = map_session_update(&session_id, &args.update);
        for event in events {
            self.sink.emit(event);
        }
        Ok(())
    }

    async fn request_permission(
        &self,
        args: RequestPermissionRequest,
    ) -> AcpResult<RequestPermissionResponse> {
        let rx = {
            let mut resolver = self.permission_resolver.lock().await;
            resolver.request(&args, self.sink.as_ref())
        };
        let response = rx.await.map_err(|e| {
            AcpError::internal_error().data(format!("Permission request channel closed: {}", e))
        })?;
        Ok(response)
    }

    async fn read_text_file(&self, args: ReadTextFileRequest) -> AcpResult<ReadTextFileResponse> {
        match tokio::fs::read_to_string(&args.path).await {
            Ok(content) => Ok(ReadTextFileResponse::new(content)),
            Err(_) => Err(AcpError::resource_not_found(Some(
                args.path.to_string_lossy().to_string(),
            ))),
        }
    }

    async fn write_text_file(
        &self,
        args: WriteTextFileRequest,
    ) -> AcpResult<WriteTextFileResponse> {
        tokio::fs::write(&args.path, &args.content)
            .await
            .map_err(|e| AcpError::internal_error().data(e.to_string()))?;
        Ok(WriteTextFileResponse::new())
    }

    async fn create_terminal(
        &self,
        args: CreateTerminalRequest,
    ) -> AcpResult<CreateTerminalResponse> {
        let terminal_id = uuid::Uuid::new_v4().to_string();

        let mut cmd = tokio::process::Command::new(&args.command);
        cmd.args(&args.args);
        if let Some(ref cwd) = args.cwd {
            cmd.current_dir(cwd);
        }
        cmd.stdin(std::process::Stdio::null());
        cmd.stdout(std::process::Stdio::piped());
        cmd.stderr(std::process::Stdio::piped());

        #[cfg(windows)]
        {
            use std::os::windows::process::CommandExt;
            cmd.creation_flags(0x08000000);
        }

        let mut child = cmd.spawn().map_err(|e| {
            AcpError::internal_error().data(format!("Failed to spawn terminal: {}", e))
        })?;

        let output = Arc::new(Mutex::new(String::new()));
        let exit_result = Arc::new(Mutex::new(None));

        if let Some(stdout) = child.stdout.take() {
            let output_clone = output.clone();
            tokio::spawn(async move {
                let mut reader = tokio::io::BufReader::new(stdout);
                let mut buf = [0u8; 4096];
                loop {
                    match reader.read(&mut buf).await {
                        Ok(0) => break,
                        Ok(n) => {
                            let text = String::from_utf8_lossy(&buf[..n]);
                            output_clone.lock().await.push_str(&text);
                        }
                        Err(_) => break,
                    }
                }
            });
        }

        if let Some(stderr) = child.stderr.take() {
            let output_clone = output.clone();
            tokio::spawn(async move {
                let mut reader = tokio::io::BufReader::new(stderr);
                let mut buf = [0u8; 4096];
                loop {
                    match reader.read(&mut buf).await {
                        Ok(0) => break,
                        Ok(n) => {
                            let text = String::from_utf8_lossy(&buf[..n]);
                            output_clone.lock().await.push_str(&text);
                        }
                        Err(_) => break,
                    }
                }
            });
        }

        let entry = TerminalEntry {
            child,
            output,
            exit_result,
        };

        self.terminals
            .lock()
            .await
            .insert(terminal_id.clone(), entry);

        Ok(CreateTerminalResponse::new(terminal_id))
    }

    async fn terminal_output(
        &self,
        args: TerminalOutputRequest,
    ) -> AcpResult<TerminalOutputResponse> {
        let tid = args.terminal_id.to_string();
        let terminals = self.terminals.lock().await;
        let entry = terminals
            .get(&tid)
            .ok_or_else(|| AcpError::resource_not_found(Some(tid.clone())))?;

        let output = entry.output.lock().await.clone();
        let exit = entry.exit_result.lock().await.clone();

        let mut response = TerminalOutputResponse::new(output, false);
        if let Some(exit_result) = exit {
            let mut exit_status = TerminalExitStatus::new();
            if let Some(code) = exit_result.code {
                exit_status = exit_status.exit_code(code as u32);
            }
            response = response.exit_status(exit_status);
        }

        Ok(response)
    }

    async fn wait_for_terminal_exit(
        &self,
        args: WaitForTerminalExitRequest,
    ) -> AcpResult<WaitForTerminalExitResponse> {
        let tid = args.terminal_id.to_string();

        // Check if already exited
        {
            let terminals = self.terminals.lock().await;
            let entry = terminals
                .get(&tid)
                .ok_or_else(|| AcpError::resource_not_found(Some(tid.clone())))?;

            let existing = entry.exit_result.lock().await;
            if let Some(ref result) = *existing {
                let mut exit_status = TerminalExitStatus::new();
                if let Some(code) = result.code {
                    exit_status = exit_status.exit_code(code as u32);
                }
                return Ok(WaitForTerminalExitResponse::new(exit_status));
            }
        }

        // Wait for child
        let exit_result_arc;
        {
            let mut terminals = self.terminals.lock().await;
            let entry = terminals
                .get_mut(&tid)
                .ok_or_else(|| AcpError::resource_not_found(Some(tid.clone())))?;

            exit_result_arc = entry.exit_result.clone();
            let status = entry.child.wait().await;
            let result = match status {
                Ok(s) => ExitResult { code: s.code() },
                Err(_) => ExitResult { code: None },
            };
            *exit_result_arc.lock().await = Some(result);
        }

        let result = exit_result_arc.lock().await;
        let mut exit_status = TerminalExitStatus::new();
        if let Some(ref r) = *result {
            if let Some(code) = r.code {
                exit_status = exit_status.exit_code(code as u32);
            }
        }
        Ok(WaitForTerminalExitResponse::new(exit_status))
    }

    async fn kill_terminal(&self, args: KillTerminalRequest) -> AcpResult<KillTerminalResponse> {
        let tid = args.terminal_id.to_string();
        let mut terminals = self.terminals.lock().await;
        let entry = terminals
            .get_mut(&tid)
            .ok_or_else(|| AcpError::resource_not_found(Some(tid)))?;

        let _ = entry.child.kill().await;
        Ok(KillTerminalResponse::new())
    }

    async fn release_terminal(
        &self,
        args: ReleaseTerminalRequest,
    ) -> AcpResult<ReleaseTerminalResponse> {
        let tid = args.terminal_id.to_string();
        let mut terminals = self.terminals.lock().await;
        if let Some(mut entry) = terminals.remove(&tid) {
            let _ = entry.child.kill().await;
        }
        Ok(ReleaseTerminalResponse::new())
    }
}
